File: entrysafe-backend/flutter-client/entry_safe_ai_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

/// AI Service for Entry Safe Apps
/// 
/// This service handles all AI-related API calls to the Entry Safe backend.
/// Each app (Accounting, Docs, Pricing) has its own endpoint.
class EntrySafeAIService {
  // TODO: Replace with your actual backend URL
  static const String _baseUrl = 'https://your-backend-url.com';
  
  /// Generate AI response for Entry Safe Accounting app
  /// 
  /// Example use cases:
  /// - "Analyze my expenses for this month"
  /// - "Generate invoice description for web development services"
  /// - "Suggest expense categories for R2500 restaurant bill"
  static Future<AIResponse> generateAccountingResponse(
    String prompt, {
    int maxTokens = 500,
    double temperature = 0.7,
  }) async {
    return _makeAIRequest(
      endpoint: '/api/ai/accounting',
      prompt: prompt,
      maxTokens: maxTokens,
      temperature: temperature,
    );
  }

  /// Generate AI response for Entry Safe Docs app
  /// 
  /// Example use cases:
  /// - "Summarize this document: [text]"
  /// - "Suggest tags for this invoice document"
  /// - "Extract key information from this receipt"
  static Future<AIResponse> generateDocsResponse(
    String prompt, {
    int maxTokens = 500,
    double temperature = 0.7,
  }) async {
    return _makeAIRequest(
      endpoint: '/api/ai/docs',
      prompt: prompt,
      maxTokens: maxTokens,
      temperature: temperature,
    );
  }

  /// Generate AI response for Entry Safe Pricing app
  /// 
  /// Example use cases:
  /// - "Suggest pricing for web design service"
  /// - "Calculate profit margin for R500 cost and R800 price"
  /// - "Competitive pricing analysis for accounting services"
  static Future<AIResponse> generatePricingResponse(
    String prompt, {
    int maxTokens = 500,
    double temperature = 0.7,
  }) async {
    return _makeAIRequest(
      endpoint: '/api/ai/pricing',
      prompt: prompt,
      maxTokens: maxTokens,
      temperature: temperature,
    );
  }

  /// Internal method to make AI API requests
  static Future<AIResponse> _makeAIRequest({
    required String endpoint,
    required String prompt,
    int maxTokens = 500,
    double temperature = 0.7,
  }) async {
    try {
      final url = Uri.parse('$_baseUrl$endpoint');
      
      final response = await http.post(
        url,
        headers: {
          'Content-Type': 'application/json',
        },
        body: jsonEncode({
          'prompt': prompt,
          'max_tokens': maxTokens,
          'temperature': temperature,
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return AIResponse.fromJson(data);
      } else {
        throw AIException(
          'Request failed with status ${response.statusCode}: ${response.body}',
        );
      }
    } catch (e) {
      throw AIException('Failed to generate AI response: $e');
    }
  }

  /// Check if AI services are available
  static Future<bool> isAIAvailable() async {
    try {
      final url = Uri.parse('$_baseUrl/api/ai/health');
      final response = await http.get(url);
      
      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['all_ready'] ?? false;
      }
      return false;
    } catch (e) {
      return false;
    }
  }
}

/// AI Response model
class AIResponse {
  final String result;
  final int tokensUsed;

  AIResponse({
    required this.result,
    required this.tokensUsed,
  });

  factory AIResponse.fromJson(Map<String, dynamic> json) {
    return AIResponse(
      result: json['result'] ?? '',
      tokensUsed: json['tokens_used'] ?? 0,
    );
  }
}

/// AI Exception
class AIException implements Exception {
  final String message;
  AIException(this.message);

  @override
  String toString() => 'AIException: $message';
}

File: entrysafe-backend/windows-client/EntrySafeAIService.cs
using System;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace EntrySafe.Services
{
    /// <summary>
    /// AI Service for Entry Safe Windows Apps
    /// Handles all AI-related API calls to the Entry Safe backend
    /// </summary>
    public class EntrySafeAIService
    {
        private static readonly HttpClient _httpClient = new HttpClient();
        private const string BaseUrl = "https://your-backend-url.com"; // TODO: Update with your actual URL

        /// <summary>
        /// Generate AI response for Entry Safe Accounting app
        /// </summary>
        public static async Task<AIResponse> GenerateAccountingResponseAsync(
            string prompt,
            int maxTokens = 500,
            double temperature = 0.7)
        {
            return await MakeAIRequestAsync(
                endpoint: "/api/ai/accounting",
                prompt: prompt,
                maxTokens: maxTokens,
                temperature: temperature
            );
        }

        /// <summary>
        /// Generate AI response for Entry Safe Docs app
        /// </summary>
        public static async Task<AIResponse> GenerateDocsResponseAsync(
            string prompt,
            int maxTokens = 500,
            double temperature = 0.7)
        {
            return await MakeAIRequestAsync(
                endpoint: "/api/ai/docs",
                prompt: prompt,
                maxTokens: maxTokens,
                temperature: temperature
            );
        }

        /// <summary>
        /// Generate AI response for Entry Safe Pricing app
        /// </summary>
        public static async Task<AIResponse> GeneratePricingResponseAsync(
            string prompt,
            int maxTokens = 500,
            double temperature = 0.7)
        {
            return await MakeAIRequestAsync(
                endpoint: "/api/ai/pricing",
                prompt: prompt,
                maxTokens: maxTokens,
                temperature: temperature
            );
        }

        /// <summary>
        /// Internal method to make AI API requests
        /// </summary>
        private static async Task<AIResponse> MakeAIRequestAsync(
            string endpoint,
            string prompt,
            int maxTokens,
            double temperature)
        {
            try
            {
                var requestBody = new
                {
                    prompt = prompt,
                    max_tokens = maxTokens,
                    temperature = temperature
                };

                var json = JsonSerializer.Serialize(requestBody);
                var content = new StringContent(json, Encoding.UTF8, "application/json");

                var response = await _httpClient.PostAsync($"{BaseUrl}{endpoint}", content);

                if (response.IsSuccessStatusCode)
                {
                    var responseString = await response.Content.ReadAsStringAsync();
                    var aiResponse = JsonSerializer.Deserialize<AIResponse>(responseString, new JsonSerializerOptions
                    {
                        PropertyNameCaseInsensitive = true
                    });

                    return aiResponse ?? throw new Exception("Failed to deserialize response");
                }
                else
                {
                    var errorContent = await response.Content.ReadAsStringAsync();
                    throw new AIException($"Request failed with status {response.StatusCode}: {errorContent}");
                }
            }
            catch (Exception ex)
            {
                throw new AIException($"Failed to generate AI response: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Check if AI services are available
        /// </summary>
        public static async Task<bool> IsAIAvailableAsync()
        {
            try
            {
                var response = await _httpClient.GetAsync($"{BaseUrl}/api/ai/health");
                
                if (response.IsSuccessStatusCode)
                {
                    var responseString = await response.Content.ReadAsStringAsync();
                    var healthCheck = JsonSerializer.Deserialize<AIHealthCheck>(responseString, new JsonSerializerOptions
                    {
                        PropertyNameCaseInsensitive = true
                    });

                    return healthCheck?.AllReady ?? false;
                }
                
                return false;
            }
            catch
            {
                return false;
            }
        }
    }

    /// <summary>
    /// AI Response model
    /// </summary>
    public class AIResponse
    {
        public string Result { get; set; } = string.Empty;
        public int TokensUsed { get; set; }
    }

    /// <summary>
    /// AI Health Check model
    /// </summary>
    public class AIHealthCheck
    {
        public bool AllReady { get; set; }
    }

    /// <summary>
    /// AI Exception
    /// </summary>
    public class AIException : Exception
    {
        public AIException(string message) : base(message) { }
        public AIException(string message, Exception innerException) : base(message, innerException) { }
    }
}
